packages/next/src/client/components/router-reducer/refetch-inactive-parallel-segments.ts

```diff
@@ -0,0 +1,89 @@
+import type { FlightRouterState } from '../../../server/app-render/types'
+import type { CacheNode } from '../../../shared/lib/app-router-context.shared-runtime'
+import type { AppRouterState } from './router-reducer-types'
+import { createEmptyCacheNode } from '../app-router'
+import { applyFlightData } from './apply-flight-data'
+import { fetchServerResponse } from './fetch-server-response'
+import { PAGE_SEGMENT_KEY } from '../../../shared/lib/segment'
+
+/**
+ * Refreshes inactive segments that are still in the current FlightRouterState.
+ * A segment is considered "inactive" when the server response indicates it didn't match to a page component.
+ * This happens during a soft-navigation, where the server will want to patch in the segment
+ * with the "default" component, but we explicitly ignore the server in this case
+ * and keep the existing state for that segment. New data for inactive segments are inherently
+ * not part of the server response when we patch the tree, because they were associated with a response
+ * from an earlier navigation/request. For each segment, once it becomes "active", we encode the URL that provided
+ * the data for it. This function traverses parallel routes looking for these markers so that it can re-fetch
+ * and patch the new data into the tree.
+ */
+export async function refreshInactiveParallelSegments({
+  state,
+  newTree,
+  newCache,
+  includeNextUrl,
+  clearExistingCache,
+}: {
+  state: AppRouterState
+  newTree: FlightRouterState
+  newCache: CacheNode
+  includeNextUrl: boolean
+  clearExistingCache: boolean
+}) {
+  const [, parallelRoutes, refetchUrl, refetchMarker] = newTree
+
+  if (
+    refetchUrl &&
+    refetchUrl !== state.canonicalUrl &&
+    refetchMarker === 'refetch'
+  ) {
+    const fetchResponse = await fetchServerResponse(
```

Comment:
we're recursing into child parallel segments below, so wouldn't this issue multiple fetches if we have multiple inactive parallel segments? feels like it'd be nice if we could just do it in just one... unless there's something i'm missing here?

Reply:
Yeah, part of this PR still being a draft will be to optimize this part a bit further!